src/main/java/com/commercetools/sync/customobjects/CustomObjectSync.java

```diff
@@ -0,0 +1,332 @@
+     * @param customObjectService the custom object service which is responsible for fetching/caching the
+     */
+    CustomObjectSync(
+        @Nonnull final CustomObjectSyncOptions syncOptions,
+        @Nonnull final CustomObjectService customObjectService) {
+
+        super(new CustomObjectSyncStatistics(), syncOptions);
+        this.customObjectService = customObjectService;
+    }
+
+    /**
+     * Iterates through the whole {@code customObjectDrafts} list and accumulates its valid drafts to batches.
+     * Every batch is then processed by {@link CustomObjectSync#processBatch(List)}.
+     *
+     * <p><strong>Inherited doc:</strong>
+     * {@inheritDoc}
+     *
+     * @param customObjectDrafts {@link List} of {@link CustomObjectDraft}'s that would be synced into CTP project.
+     * @return {@link CompletionStage} with {@link CustomObjectSyncStatistics} holding statistics of all sync
+     *     processes performed by this sync instance.
+     */
+    protected CompletionStage<CustomObjectSyncStatistics> process(
+        @Nonnull final List<CustomObjectDraft<JsonNode>> customObjectDrafts) {
+        final List<List<CustomObjectDraft<JsonNode>>> batches = batchElements(
+            customObjectDrafts, syncOptions.getBatchSize());
+        return syncBatches(batches, CompletableFuture.completedFuture(statistics));
+    }
+
+    /**
+     * This method first creates a new {@link Set} of valid {@link CustomObjectDraft} elements. For more on the rules of
+     * validation, check: {@link CustomObjectSync#validateDraft(CustomObjectDraft)}. Using the resulting set of
+     * {@code validCustomObjectDrafts}, the matching custom objects in the target CTP project are fetched then the
+     * method {@link CustomObjectSync#syncBatch(Set, Set)} is called to perform the sync (<b>update</b> or <b>create</b>
+     * requests accordingly) on the target project.
+     *
+     * <p> In case of error during of fetching of existing custom objects, the error callback will be triggered.
+     * And the sync process would stop for the given batch.
+     * </p>
+     *
+     * @param batch batch of drafts that need to be synced
+     * @return a {@link CompletionStage} containing an instance
+     *      of {@link CustomObjectSyncStatistics} which contains information about the result of syncing the supplied
+     *      batch to the target project.
+     */
+    protected CompletionStage<CustomObjectSyncStatistics> processBatch(
+        @Nonnull final List<CustomObjectDraft<JsonNode>> batch) {
+
+        final Set<CustomObjectDraft<JsonNode>> validCustomObjectDrafts = batch.stream().filter(
+            this::validateDraft).collect(toSet());
+
+        if (validCustomObjectDrafts.isEmpty()) {
+            statistics.incrementProcessed(batch.size());
+            return completedFuture(statistics);
+        } else {
+            final Set<CustomObjectCompositeIdentifier> identifiers = validCustomObjectDrafts.stream().map(
```

Comment:
I am afraid that it may not help because `CustomObjectCompositeIdentifier` instantiation is a static method, which would be shared in multithread environment.